source/simulator/stokes_matrix_free.cc

```diff
@@ -0,0 +1,1223 @@
+                 "GridGenerator::..., setting colorize=true during mesh generation "
+                 "and calling make_no_normal_flux_constraints() for each no normal "
+                 "flux boundary will fulfill the condition."));
+
+        mg_constrained_dofs.make_zero_boundary_constraints(dof, bid_set, comp_mask);
+      }
+    }
+
+
+    namespace ChangeVectorTypes
+    {
+      void import(TrilinosWrappers::MPI::Vector &out,
+                  const dealii::LinearAlgebra::ReadWriteVector<double> &rwv,
+                  const VectorOperation::values                 operation)
+      {
+        Assert(out.size() == rwv.size(),
+               ExcMessage("Both vectors need to have the same size for import() to work!"));
+
+        Assert(out.locally_owned_elements() == rwv.get_stored_elements(),
+               ExcNotImplemented());
+
+        if (operation == VectorOperation::insert)
+          {
+            for (const auto idx : out.locally_owned_elements())
+              out[idx] = rwv[idx];
+          }
+        else if (operation == VectorOperation::add)
+          {
+            for (const auto idx : out.locally_owned_elements())
+              out[idx] += rwv[idx];
+          }
+        else
+          AssertThrow(false, ExcNotImplemented());
+
+        out.compress(operation);
+      }
+
+
+      void copy(TrilinosWrappers::MPI::Vector &out,
+                const dealii::LinearAlgebra::distributed::Vector<double> &in)
+      {
+        dealii::LinearAlgebra::ReadWriteVector<double> rwv(out.locally_owned_elements());
+        rwv.import(in, VectorOperation::insert);
+        //This import function doesn't exist until after dealii 9.0
+        //Implemented above
+        import(out, rwv,VectorOperation::insert);
+      }
+
+      void copy(dealii::LinearAlgebra::distributed::Vector<double> &out,
+                const TrilinosWrappers::MPI::Vector &in)
+      {
+        dealii::LinearAlgebra::ReadWriteVector<double> rwv;
+        rwv.reinit(in);
+        out.import(rwv, VectorOperation::insert);
+      }
+
+      void copy(TrilinosWrappers::MPI::BlockVector &out,
+                const dealii::LinearAlgebra::distributed::BlockVector<double> &in)
+      {
+        const unsigned int n_blocks = in.n_blocks();
+        for (unsigned int b=0; b<n_blocks; ++b)
+          copy(out.block(b),in.block(b));
+      }
+
+      void copy(dealii::LinearAlgebra::distributed::BlockVector<double> &out,
+                const TrilinosWrappers::MPI::BlockVector &in)
+      {
+        const unsigned int n_blocks = in.n_blocks();
+        for (unsigned int b=0; b<n_blocks; ++b)
+          copy(out.block(b),in.block(b));
+      }
+    }
+
+
+    /**
+             * Implement the block Schur preconditioner for the Stokes system.
+             */
+    template <class ABlockMatrixType, class StokesMatrixType, class MassMatrixType, class PreconditionerMp,class PreconditionerA>
+    class BlockSchurPreconditioner : public Subscriptor
```

Comment:
This has the same name as the one in solver.cc. What is the precise difference, just template arguments for the types?
If you think we need to keep both, can you rename it? Otherwise a unity build fails: https://jenkins.geodynamics.org/blue/organizations/jenkins/aspect/detail/PR-2987/8/pipeline/#step-45-log-730

Reply:
The main difference is that
{
        dealii::LinearAlgebra::distributed::BlockVector<double>  dst_tmp(dst);
        dst_tmp.block(0) = 0.0;
        stokes_matrix.vmult(utmp, dst_tmp); // B^T
        utmp.block(0) *= -1.0;
        utmp.block(0) += src.block(0);
}
Is how we have to compute `B^T`. The other just does `stokes_matrix.block(0,1).vmult()`

Should we just rename it `BlockSchurGMGPreconditioner`?

Reply:
you need to give it a different name, otherwise unity builds fail (that is the reason the tester is unhappy)